packages/ceramic-core/src/document.ts

```diff
@@ -274,35 +290,77 @@ class Document extends EventEmitter {
   }
 
   async _fetchLog (cid: CID, log: Array<CID> = []): Promise<Array<CID>> {
-    if (this._doctype.state.log.some(x => x.equals(cid))) { // already processed
+    if (await this._isCidIncluded(cid, this._doctype.state.log)) { // already processed
       return []
     }
     const record = await this.dispatcher.retrieveRecord(cid)
-    const prevCid: CID = record.prev
+    let payload = record
+    if (DoctypeUtils.isSignedRecord(record)) {
+      payload = await this.dispatcher.retrieveRecord(record.link)
+    }
+    const prevCid: CID = payload.prev
     if (!prevCid) { // this is a fake log
       return []
     }
     log.unshift(cid)
-    if (this._doctype.state.log.some(x => x.equals(prevCid))) {
+    if (await this._isCidIncluded(prevCid, this._doctype.state.log)) {
       // we found the connection to the canonical log
       return log
     }
     return this._fetchLog(prevCid, log)
   }
 
+  /**
+   * Find index of the record in the array. If the record is signed, fetch the payload
+   * @param cid - CID value
+   * @param log - Log array
+   * @private
+   */
+  async _findIndex(cid: CID, log: Array<CID>): Promise<number> {
+    // const conflictIdx = this._doctype.state.log.findIndex(x => x.equals(record.prev)) + 1
+    for (let index = 0; index < log.length; index++) {
+      const c = log[index]
+      if (c.equals(cid)) {
+        return index
+      }
+      const record = await this.dispatcher.retrieveRecord(c)
+      if (DoctypeUtils.isSignedRecord(record) && record.link.equals(cid)) {
+        return index
+      }
+    }
+    return -1
+  }
+
+  /**
+   * Is CID included in the log. If the record is signed, fetch the payload
+   * @param cid - CID value
+   * @param log - Log array
+   * @private
+   */
+  async _isCidIncluded(cid: CID, log: Array<CID>): Promise<boolean> {
+    return (await this._findIndex(cid, log)) !== -1
+  }
+
   async _applyLog (log: Array<CID>): Promise<boolean> {
     let modified = false
-    if (log[log.length - 1].equals(this.head)) return // log already applied
+    if (log[log.length - 1].equals(this.head)) {
+      // log already applied
+      return
+    }
     const cid = log[0]
     const record = await this.dispatcher.retrieveRecord(cid)
-    if (record.prev.equals(this.head)) {
+    let payload = record
+    if (DoctypeUtils.isSignedRecord(record)) {
+      payload = await this.dispatcher.retrieveRecord(record.link)
+    }
+    if (payload.prev.equals(this.head)) {
       // the new log starts where the previous one ended
       this._doctype.state = await this._applyLogToState(log, cloneDeep(this._doctype.state))
       modified = true
     } else {
       // we have a conflict since prev is in the log of the
       // local state, but isn't the head
-      const conflictIdx = this._doctype.state.log.findIndex(x => x.equals(record.prev)) + 1
+      const conflictIdx = await this._findIndex(payload.prev, this._doctype.state.log) + 1
```

Comment:
If we have a separate `findIndex` method we might as well do the `+ 1` there as well :)

Reply:
`_findIndex` method would not be `findIndex` anymore :) Also, `_findIndex` is used in other places as well

Reply:
oh, ok :)

Reply:
👍